linera-core/src/worker.rs

```diff
@@ -801,29 +836,29 @@ where
         chain_id: ChainId,
         message_id: MessageId,
     ) -> Result<Option<IncomingMessage>, WorkerError> {
-        let Some(certificate) = self
-            .read_certificate(message_id.chain_id, message_id.height)
-            .await?
-        else {
+        let sender = message_id.chain_id;
+        let index = usize::try_from(message_id.index).map_err(|_| ArithmeticError::Overflow)?;
+        let Some(certificate) = self.read_certificate(sender, message_id.height).await? else {
             return Ok(None);
         };
-
-        let index = usize::try_from(message_id.index).map_err(|_| ArithmeticError::Overflow)?;
-        let Some(outgoing_message) = certificate.value().messages().get(index).cloned() else {
+        let Some(messages) = certificate.value().messages() else {
+            return Ok(None);
+        };
+        let Some(outgoing_message) = messages.get(index).cloned() else {
             return Ok(None);
         };
 
-        let application_id = outgoing_message.message.application_id();
-        let origin = Origin {
-            sender: message_id.chain_id,
-            medium: match outgoing_message.destination {
-                Destination::Recipient(_) => Medium::Direct,
-                Destination::Subscribers(name) => Medium::Channel(ChannelFullName {
+        let medium = match outgoing_message.destination {
+            Destination::Recipient(_) => Medium::Direct,
+            Destination::Subscribers(name) => {
+                let application_id = outgoing_message.message.application_id();
+                Medium::Channel(ChannelFullName {
                     application_id,
                     name,
-                }),
-            },
+                })
+            }
         };
+        let origin = Origin { sender, medium };
```

Comment:
It's a bit hard to parse the lines but this seems like a vanilla code refactoring, right? Consider a separate commit next time to speed up reviewing.

Reply:
Right, sorry.
It's mostly just refactoring, but also dealing with `messages` now returning an option.
It got rebased, squashed and split up a few times already, and I failed to properly separate the pure refactoring part in the end. :sweat_smile: